### ISSUE TITLE
Provide API to create mock object on the fly

### ISSUE THREAD

**Issue by benxshen** [jmockit/jmockit1#250]
I saw  #232 addressed the same request, but I can't add my comment since it was locked, and only to open a new issue to discuss again.

Said there is a businessService to provide some business operations, and will send notification mail to users before end of operation. I would have integration tests to test all my Spring beans and configurations (by @RunWith(SpringJUnit4ClassRunner.class) annotated on Test Class), but I want my MailSender mocked only (to make sure sendMail() is invoked but not to send real mail out). In this situation, I need to create an mocked object on the fly in the @Configuration class to provide the mocked bean to override actual one.

I tried to find solutions on the web, a nearly workaround is found, which uses the Deencapsulation.setField(myService,  "mailSender", mockedMailSender) in the @Before method of Test Class. It works for my tests, but I don't know if I make use of Deencapsulation.setField correctly. And more, "Deencapsulation.setField" is not good semantic when putting it in the Test. (It could be my personal issue on the Deencapsulation understanding.)

Another approach I can think (but I don't know how it's possible) is to auto inject @Injectable  fields to @Autoware/@Resource fields, in this way, I don't need to manually create mocked object on the fly, and JMockit mixes in the @Injectable  fields to my Spring managed bean for integration tests.

Last, I appreciate and feel awesome to the power of JMockit, thank you.

**Comment by rliesenfeld** [comment]
What about the example solution shown in the following test class:

https://github.com/jmockit/jmockit1/blob/2142240544f44394900646236224afc5fbeb7604/samples/mockito/test/mockmvc/example/WebAppContextJMockitTest.java

It uses SpringJUnit4ClassRunner and a @Configuration, with a couple of mocked beans. Should be applicable to your case, I think.

**Comment by benxshen** [comment]
Thank you, rliesenfeld :)
Yes, this example is similar to my case. I use SpringJUnit4ClassRunner to run my tests and  @Configuration inner class to provide mocked beans for testing. 

As per your example, it's possible to use @Captuing to provide mocked beans for testing. I tested the same, and it works good. No doubt, @Capturing is a powerful tool and it's more easier for us to write tests. The only drawback is the code (in the Config class) is not much clear to show a mocked bean is created here. A mock() method is more clearly. :p

BTW, another question is that I can't pass tests  when I played the JMockit with List/Map by using Mocked, Injectable, and Capturing. Are they not supported in JMockit to be mocked?

Here is a simple test:

public class JMockitCapturingTest {
	@Capturing
	List<String> anyListWillBeCapturedAsMockedList;

	@Capturing
	Map<String, Object> anyMapWillBeCapturedAsMockedMap;


	@Test
	public void testCapturing() throws Exception {

		List<String> anotherMockedList = new ArrayList<String>();
		System.out.println(anotherMockedList);

		Map<String, Object> anotherMockedMap = new HashMap<String, Object>();
		System.out.println(anotherMockedMap);

		new Expectations() {{
			anotherMockedList.get(0); result = "hello";
			anotherMockedMap.get("hello"); result = "world";
		}};


		assertEquals("hello", anotherMockedList.get(0));
		assertEquals("world", anotherMockedMap.get("hello"));
	}

}

**Comment by rliesenfeld** [comment]
Ok, so use of @Capturing solves the issue.

As for mocking JRE classes like ArrayList and HashMap, yes, there are restrictions. However, mocking such classes (or just creating mock Lists and Maps) is generally regarded as something to avoid. Besides, it's difficult to see a real-world situation where such mocking is needed or desirable.

**Comment by Davidside** [comment]
Yeah I could solve this by @capturing annotation, but it has two down sides.

Firstly I dont want to create an anonnymous class of interface with empty methods in context config, because intefaces has a lots of methods. So I can use MockUp() {}.getMockInstance(); instead, but this sometimes fails on NPE for some services. But this could be handled with different own solution like creating a mock from interface, the same way how EasyMock or Mockito do it.

Caused by: java.lang.NullPointerException
at com.sun.beans.TypeResolver.resolve(TypeResolver.java:203)
at com.sun.beans.TypeResolver.resolve(TypeResolver.java:218)
at com.sun.beans.TypeResolver.resolveInClass(TypeResolver.java:96)
at java.beans.FeatureDescriptor.getParameterTypes(FeatureDescriptor.java:387)
at java.beans.MethodDescriptor.setMethod(MethodDescriptor.java:114)
at java.beans.MethodDescriptor.(MethodDescriptor.java:72)
at java.beans.MethodDescriptor.(MethodDescriptor.java:56)
at java.beans.Introspector.getTargetMethodInfo(Introspector.java:1149)
at java.beans.Introspector.getBeanInfo(Introspector.java:416)
at java.beans.Introspector.getBeanInfo(Introspector.java:163)
at org.springframework.beans.CachedIntrospectionResults.(CachedIntrospectionResults.java:224)
at org.springframework.beans.CachedIntrospectionResults.forClass(CachedIntrospectionResults.java:149)
at org.springframework.beans.BeanWrapperImpl.getCachedIntrospectionResults(BeanWrapperImpl.java:324)
at org.springframework.beans.BeanWrapperImpl.getPropertyDescriptors(BeanWrapperImpl.java:331)
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.filterPropertyDescriptorsForDependencyCheck(AbstractAutowireCapableBeanFactory.java:1242)
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1101)
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:517)
... 36 more

Secondly (the bigger issue), when I use Capturing, running the test takes inadequate amount of time and it also grows linearly with every other Capturing annotation. It means when I dont use Capturing, it takes about a 1 second to run a test, when I use one Capturing annotation, it takes 7 seconds, when I use two Capturing it takes 14 seconds and so on. I need to write an integration test with 7 services, so it is not acceptable to take it so much time.

**Comment by rliesenfeld** [comment]
@Davidside No need to "create an anonnymous class of interface", or use any mock-up; just use `@Capturing`, as shown in the example tests of the comment of 24 Jan.

About bad performance with `@Capturing`, I would need an example test, since JMockit's own tests which use this annotation are not displaying the problem.

**Comment by Davidside** [comment]
@rliesenfeld Yeah it works with null value for beans. Thanks for advice.

About performance, below I wrote an example test code and also I attached tree call from JProfiler. It seems it has something to do with WicketTester. The attached tree call is for run with seven Capturing fields in test. It grows linearly with every other Capturing field, also with every use of wicketTester.startComponent().

`@RunWith(SpringJUnit4ClassRunner.class)`
`@ContextConfiguration(loader=AnnotationConfigContextLoader.class)`
`public class TestCase {`

	private WicketTester wicketTester;
	
    @Configuration
    static class ContextConfiguration {
        @Bean(name="globalCache") public ICacheHelper globalCache() { return null; }
    }
	
	@Capturing ICacheHelper globalCache;
	
	@Test
	public void testName() throws Exception {
		wicketTester = new WicketTester();
	}
`}`
![jmockit_capturing_performance](https://user-images.githubusercontent.com/17691090/29271554-58ef2072-80fc-11e7-9277-99a3441ba94d.PNG)
